openpype/tools/ayon_loader/models/actions.py

```diff
@@ -0,0 +1,870 @@
+        }
+
+        folder_ids = {p["parent"] for p in product_docs_by_id.values()}
+        folder_docs = get_assets(project_name, asset_ids=folder_ids)
+        folder_docs_by_id = {
+            f["_id"]: f for f in folder_docs
+        }
+
+        project_doc = get_project(project_name)
+        project_doc["code"] = project_doc["data"]["code"]
+
+        for product_id, product_doc in product_docs_by_id.items():
+            folder_id = product_doc["parent"]
+            folder_doc = folder_docs_by_id[folder_id]
+            product_context_by_id[product_id] = {
+                "project": project_doc,
+                "asset": folder_doc,
+                "subset": product_doc,
+            }
+
+        for repre_doc in repre_docs:
+            version_id = repre_doc["parent"]
+            version_doc = version_docs_by_id[version_id]
+            product_id = version_doc["parent"]
+            product_doc = product_docs_by_id[product_id]
+            folder_id = product_doc["parent"]
+            folder_doc = folder_docs_by_id[folder_id]
+
+            repre_context_by_id[repre_doc["_id"]] = {
+                "project": project_doc,
+                "asset": folder_doc,
+                "subset": product_doc,
+                "version": version_doc,
+                "representation": repre_doc,
+            }
+        return product_context_by_id, repre_context_by_id
+
+    def _get_action_items_for_contexts(
+        self,
+        project_name,
+        version_context_by_id,
+        repre_context_by_id
+    ):
+        """Prepare action items based on contexts.
+
+        Actions are prepared based on discovered loader plugins and contexts.
+        The context must be valid for the loader plugin.
+
+        Args:
+            project_name (str): Project name.
+            version_context_by_id (dict[str, dict[str, Any]]): Version
+                contexts by version id.
+            repre_context_by_id (dict[str, dict[str, Any]]): Representation
+        """
+
+        action_items = []
+        if not version_context_by_id and not repre_context_by_id:
+            return action_items
+
+        product_loaders, repre_loaders = self._get_loaders(project_name)
+
+        repre_contexts_by_name = collections.defaultdict(list)
+        for repre_context in repre_context_by_id.values():
+            repre_name = repre_context["representation"]["name"]
+            repre_contexts_by_name[repre_name].append(repre_context)
+
+        for loader in repre_loaders:
+            # # do not allow download whole repre, select specific repre
+            # if tools_lib.is_sync_loader(loader):
+            #     continue
+
+            for repre_name, repre_contexts in repre_contexts_by_name.items():
+                filtered_repre_contexts = filter_repre_contexts_by_loader(
+                    repre_contexts, loader)
+                if len(filtered_repre_contexts) != len(repre_contexts):
+                    continue
```

Comment:
Are we sure this is correct - technically a representation that matches _only by name_ might have on representation be valid for a loader and another won't.

There could be two representations named `ma` but the loader could be filtered out by `family`
There could be two representations named `render` but the loader could be filtered out by `extensions`

I would expect the continue statement to be:

```suggestion
                if not filtered_repre_contexts:
                    continue
```

Or am I misreading this code?

Reply:
You're right, I don't know why it was there TBH.